dnf/transaction_sr.py

```diff
@@ -320,7 +323,7 @@ def _replay_pkg_action(self, pkg_data):
             self._nevra_reason_cache[nevra] = reason
 
         if action in ("Install", "Upgrade", "Downgrade"):
-            if action == "Install" and query_na.installed() and not self._base._get_installonly_query(query_na):
+            if action == "Install" and query_na.installed() and not self._base._get_installonly_query(query_na) and not self._history_mode:
                 self._raise_or_warn(self._ignore_installed,
                     _('Package "{na}" is already installed for action "{action}".').format(na=na, action=action))
                 return
```

Comment:
Looking at what behavior you're trying to change with the `history_mode` and I actually think the `return` on this line shouldn't be here.

In case we're ignoring installed packages, we should print the warning but actually install the requested version, right? If we're not ignoring installed packages, exception is raised and the execution stops here.

If we remove the `return`, then the only difference between having `history_mode` and not is the warning being printed.

Reply:
Removed `return` and wrote a `history redo` test that turns an install into a downgrade.